reference-library/TransientStateLibrary.md

```diff
@@ -0,0 +1,122 @@
+| manager | IPoolManager | The PoolManager contract instance
+| caller_ | address | The address of the caller
+| currency | Currency | The currency for which to lookup the delta             |
+
+**Returns:**
+
+- `int256`: The delta value for the specified caller and currency
+
+### isUnlocked
+
+```solidity
+function isUnlocked(IPoolManager manager) internal view returns (bool)
+```
+
+Checks if the PoolManager contract is currently unlocked.
+
+| Param Name   | Type         | Description                          |
+|--------------|--------------|--------------------------------------|
+| manager | IPoolManager | The PoolManager contract instance
+
+**Returns:**
+
+- `bool`: `true` if the contract is unlocked, `false` otherwise
+
+## Usage and Importance
+The `TransientStateLibrary` plays a critical role in Uniswap V4's operation:
+
+1. **Gas Optimization:** By using transient storage, the library helps reduce gas costs associated with state changes that are only relevant within a single transaction. This is particularly important for multi-hop transactions, where internal net balances (deltas) are updated instead of making token transfers for each hop.
+2. **Security:** The library provides functions to check the lock state and manage deltas, which are crucial for maintaining the integrity of the protocol during operations. The use of transient storage also allows for more efficient implementation of security measures compared to V3's reentrancy guards.
+3. **Flexibility:** The library allows for efficient management of currency-specific data, such as reserves and deltas, which is essential for Uniswap V4's multi-currency pools.
+4. **Encapsulation:** By centralizing these utility functions in a library, the code promotes better organization and reusability across the Uniswap V4 codebase.
+
+## Integration with PoolManager
+
+The `TransientStateLibrary` is designed to work closely with the `PoolManager` contract. All functions in this library take an `IPoolManager` instance as their first parameter, allowing them to interact with the PoolManager's transient storage.
+
+## Security Considerations
+
+When using this library, it's important to note:
+
+1. The library relies on transient storage, which is cleared at the end of each transaction. Any critical state that needs to persist across transactions should not solely rely on this mechanism.
+2. The `isUnlocked` function is crucial for ensuring that certain operations are only performed when the contract is in the correct state. Always check this before performing sensitive operations.
+3. The `currencyDelta` function returns signed integers. Be careful when performing arithmetic with these values to avoid underflows or overflows.
+
+By leveraging the `TransientStateLibrary`, Uniswap V4 achieves a balance between efficiency and security in managing its transient state, contributing to the overall robustness of the protocol.
```

Comment:
we can exclude this section -- TransientStateLibrary is practically always safe to use in an execution context